sklearn/linear_model/_sgd_fast.pyx.tp

```diff
@@ -681,15 +797,27 @@ def _plain_sgd(const double[::1] weights,
         epoch + 1
     )
 
+{{endfor}}
+
 
-cdef bint any_nonfinite(const double *w, int n) noexcept nogil:
+{{for name_suffix, c_type, np_type in dtypes}}
+
+cdef bint any_nonfinite{{name_suffix}}(const {{c_type}} *w, int n) noexcept nogil:
     for i in range(n):
-        if not skl_isfinite(w[i]):
+        if not skl_isfinite{{name_suffix}}(w[i]):
             return True
     return 0
 
+{{endfor}}
+
+
+{{for name_suffix, c_type, np_type in dtypes}}
 
-cdef double sqnorm(double * x_data_ptr, int * x_ind_ptr, int xnnz) noexcept nogil:
+cdef double sqnorm{{name_suffix}}(
```

Comment:
I think `sqnorm` can be updated to be a inline fused type and does not need to be templated:

```python
cdef inline double sqnorm(
    floating * x_data_ptr,
    int * x_ind_ptr,
    int xnnz,
) noexcept nogil:
    cdef double x_norm = 0.0
    cdef int j
    cdef double z
    for j in range(xnnz):
        z = x_data_ptr[j]
        x_norm += z * z
    return x_norm
```